### PR TITLE
gdal enable shared library and added ossim, openthreads, openjpeg

### PR DESCRIPTION
**Pull request by bugbrains**
reopening pull request - https://github.com/mxe/mxe/pull/527. this pull request incorporates all comments and modifications of previous one. Initial was dirty and contains a lot of forcing into shared library. usage of `--enable-shared --disable static` instead of `MXE_CONFIGURE_OPTS`

Only issue is downgrading of qwt_qt4 to 5.0. This is because of OTB monteverdi2 which requires 5.x version. I will try to clean it up later.

enabled netcdf, hdf4 shared. 
libgeotiff shared for x86_64. 
new package qwt5_qt4 
new package otb

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1402,6 +1402,22 @@ <h2 id="packages">List of Packages</h2>
...
+        <td class="package">ossim</td>
+        <td class="website"><a href="http://www.ossim.org/">ossim</a></td>
+    </tr>
+    <tr>
+        <td class="package">otb</td>
+        <td class="website"><a href="http://www.orfeo-toolbox.org/">Orfeo Toolbox</a></td>
+    </tr>
+    <tr>
+        <td class="package">openjpeg</td>
+        <td class="website"><a href="https://code.google.com/p/openjpeg/">ossim</a></td>
```
Comment (TimothyGu) [inline index.html]: this isn't ossim

File: index.html [hunk 2]
```diff
@@ -1402,6 +1402,22 @@ <h2 id="packages">List of Packages</h2>
         <td class="website"><a href="http://glew.sourceforge.net/">GLEW</a></td>
     </tr>
     <tr>
+        <td class="package">openthreads</td>
```
Comment (TimothyGu) [inline index.html]: ALphabetical order

File: src/tinyxml.mk [hunk 3]
```diff
@@ -17,18 +17,16 @@ define $(PKG)_UPDATE
...
-    $(TARGET)-ranlib '$(1)/libtinyxml.a'
-    $(INSTALL) -d               '$(PREFIX)/$(TARGET)/lib'
-    $(INSTALL) -m644 '$(1)'/*.a '$(PREFIX)/$(TARGET)/lib/'
-    $(INSTALL) -d               '$(PREFIX)/$(TARGET)/include'
-    $(INSTALL) -m644 '$(1)'/*.h '$(PREFIX)/$(TARGET)/include/'
+    mkdir '$(1).build'
+    cd '$(1).build' && cmake \
+        -DCMAKE_TOOLCHAIN_FILE='$(CMAKE_TOOLCHAIN_FILE)' \
+        -DBUILD_SHARED_LIBS=$(if $(BUILD_STATIC),FALSE,TRUE) \
+    -DCMAKE_CXX_FLAGS='-DTIXML_USE_STL' \
```
Comment (TimothyGu) [inline src/tinyxml.mk]: Indentation

File: src/tinyxml.mk [hunk 4]
```diff
@@ -17,18 +17,16 @@ define $(PKG)_UPDATE
...
+        '$(1)'
+    $(MAKE) -C '$(1).build' install
```
Comment (TimothyGu) [inline src/tinyxml.mk]: `-j $(JOBS)`

File: src/qwt5_qt4.mk [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+define $(PKG)_UPDATE
+    echo 'TODO: Updates for package qwt5 need to be written.' >&2;
+    echo '$(qwt5_VERSION)'
+endef
+
+define $(PKG)_BUILD
+    $(SED) -i '/\INSTALLBASE /s%\INSTALLBASE .*%INSTALLBASE=$(PREFIX)/$(TARGET)/qt%g' '$(1)/qwtconfig.pri'
+    $(if $(BUILD_STATIC),\
+        echo "QWT_CONFIG -= QwtDll" >> '$(1)/qwtconfig.pri')
+ # build
```
Comment (TimothyGu) [inline src/qwt5_qt4.mk]: Indentation

File: src/qwt5_qt4-1-exampleslib.patch [hunk 6]
```diff
@@ -0,0 +1,8 @@
+diff -bur qwt5-5.2.orig/examples/simple_plot/simple_plot.pro qwt5-5.2/examples/simple_plot/simple_plot.pro
+--- qwt5-5.2.orig/examples/simple_plot/simple_plot.pro	2014-10-23 16:47:44.952285654 +0530
++++ qwt5-5.2/examples/simple_plot/simple_plot.pro	2014-10-23 16:49:32.033763742 +0530
+@@ -14,3 +14,4 @@
+ SOURCES = \
+     simple.cpp
+ 
++LIBS+= "-lqwt5"
```
Comment (TimothyGu) [inline src/qwt5_qt4-1-exampleslib.patch]: This won't work with static builds

File: src/proj.mk [hunk 7]
```diff
@@ -19,7 +19,7 @@ endef
 
 define $(PKG)_BUILD
     cd '$(1)' && ./configure \
-        $(MXE_CONFIGURE_OPTS) \
+    $(MXE_CONFIGURE_OPTS) \
```
Comment (TimothyGu) [inline src/proj.mk]: Unrelated change

File: src/portablexdr.mk [hunk 8]
```diff
@@ -20,9 +20,8 @@ endef
 define $(PKG)_BUILD
     cd '$(1)' && ./configure \
         $(MXE_CONFIGURE_OPTS) \
-        AR='$(TARGET)-ar'
+        AR='$(TARGET)-ar' \
+        CFLAGS='-Wno-error=pointer-to-int-cast'
```
Comment (TimothyGu) [inline src/portablexdr.mk]: This is not safe at all. Find the error in the code and fix it instead of ignoring the warning/error.

File: src/otb.mk [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+define $(PKG)_BUILD
+    mkdir '$(1).build'
+#copy otb.conf to build directory.
+    cp '$(1)/otb.conf' '$(1).build/otb.conf'
+    cd '$(1).build' && cmake \
+        -DCMAKE_TOOLCHAIN_FILE='$(CMAKE_TOOLCHAIN_FILE)' \
+    $(if $(BUILD_SHARED),\
+        -DBUILD_SHARED_LIBS=TRUE ) \
+    -DBUILD_TESTING=FALSE \
```
Comment (TimothyGu) [inline src/otb.mk]: Indentation, here and below

File: src/otb.mk [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    -DBUILD_EXAMPLES=FALSE \
+    -DBUILD_APPLICATIONS=TRUE \
+    -DCMAKE_USE_PTHREADS=FALSE \
+    -DCMAKE_USE_WIN32_THREADS=FALSE \
```
Comment (TimothyGu) [inline src/otb.mk]: why not?

File: src/otb.mk [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    mkdir '$(1).build'
+#copy otb.conf to build directory.
+    cp '$(1)/otb.conf' '$(1).build/otb.conf'
+    cd '$(1).build' && cmake \
+        -DCMAKE_TOOLCHAIN_FILE='$(CMAKE_TOOLCHAIN_FILE)' \
+    $(if $(BUILD_SHARED),\
+        -DBUILD_SHARED_LIBS=TRUE ) \
+    -DBUILD_TESTING=FALSE \
+    -DBUILD_EXAMPLES=FALSE \
+    -DBUILD_APPLICATIONS=TRUE \
```
Comment (TimothyGu) [inline src/otb.mk]: we don't build applications

File: src/otb.mk [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    -DCMAKE_USE_PTHREADS=FALSE \
+    -DCMAKE_USE_WIN32_THREADS=FALSE \
+    -DGDAL_CONFIG='$(PREFIX)/$(TARGET)/bin/gdal-config' \
+    -DOSSIM_LIBRARY='$(PREFIX)/$(TARGET)/lib/libossim.dll.a;$(PREFIX)/$(TARGET)/lib/libOpenThreads.dll.a' \
+    -DOTB_COMPILE_WITH_FULL_WARNING=FALSE \
+    -DOTB_USE_PATENTED=FALSE \
```
Comment (TimothyGu) [inline src/otb.mk]: why not?

File: src/otb.mk [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    -DOTB_USE_CURL=TRUE \
```
Comment (TimothyGu) [inline src/otb.mk]: Is this using an internal copy of Curl? It shouldn't.

File: src/ossim.mk [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+define $(PKG)_UPDATE
+    $(WGET) -q -O- 'http://download.osgeo.org/ossim/source/latest/' | \
+    $(SED) -n 's,.*ossim-\([0-9][^>]*\)\.tar.*,\1,p' | \
+    head -1
+endef
+
+define $(PKG)_BUILD
+    mkdir '$(1).build'
+    cd '$(1).build' && cmake \
+    -DCMAKE_TOOLCHAIN_FILE='$(CMAKE_TOOLCHAIN_FILE)' \
```
Comment (TimothyGu) [inline src/ossim.mk]: Indentation

File: src/opencv.mk [hunk 15]
```diff
@@ -48,7 +48,7 @@ define $(PKG)_BUILD
       '$(1)'
 
     # install
-    $(MAKE) -C '$(1).build' -j '$(JOBS)' install VERBOSE=1
+    $(MAKE) -C '$(1).build' -j '$(JOBS)' install #VERBOSE=1
```
Comment (TimothyGu) [inline src/opencv.mk]: why?

File: src/opencv.mk [hunk 16]
```diff
@@ -56,8 +56,16 @@ define $(PKG)_BUILD
     $(if $(findstring x86_64-w64-mingw32,$(TARGET)),\
         $(SED) -i 's/OpenEXR//' '$(1).build/unix-install/opencv.pc')
     $(SED) -i 's,share/OpenCV/3rdparty/,,g' '$(1).build/unix-install/opencv.pc'
+    $(SED) -i 's/dll/dll.a/g' '$(1).build/unix-install/opencv.pc'
     $(INSTALL) -m755 '$(1).build/unix-install/opencv.pc' '$(PREFIX)/$(TARGET)/lib/pkgconfig'
 
+    $(if $(findstring x86_64-w64-mingw32,$(TARGET)),\
+    $(SED) -i 's/CMAKE_OPENCV_GCC_TARGET_MACHINE/OPENCV_GCC_TARGET_MACHINE/g' '$(PREFIX)/$(TARGET)/x64/mingw/lib/OpenCVConfig.cmake')
```
Comment (TimothyGu) [inline src/opencv.mk]: indentation

File: src/opencv.mk [hunk 17]
```diff
@@ -56,8 +56,16 @@ define $(PKG)_BUILD
     $(if $(findstring x86_64-w64-mingw32,$(TARGET)),\
         $(SED) -i 's/OpenEXR//' '$(1).build/unix-install/opencv.pc')
     $(SED) -i 's,share/OpenCV/3rdparty/,,g' '$(1).build/unix-install/opencv.pc'
+    $(SED) -i 's/dll/dll.a/g' '$(1).build/unix-install/opencv.pc'
     $(INSTALL) -m755 '$(1).build/unix-install/opencv.pc' '$(PREFIX)/$(TARGET)/lib/pkgconfig'
```
Comment (TimothyGu) [inline src/opencv.mk]: while at it please change permission 755 to 644

File: src/opencv.mk [hunk 18]
```diff
@@ -56,8 +56,16 @@ define $(PKG)_BUILD
...
 
+    $(if $(findstring x86_64-w64-mingw32,$(TARGET)),\
+    $(SED) -i 's/CMAKE_OPENCV_GCC_TARGET_MACHINE/OPENCV_GCC_TARGET_MACHINE/g' '$(PREFIX)/$(TARGET)/x64/mingw/lib/OpenCVConfig.cmake')
+    $(if $(findstring i686-pc-mingw32,$(TARGET)),\
+    $(SED) -i 's/CMAKE_OPENCV_GCC_TARGET_MACHINE/OPENCV_GCC_TARGET_MACHINE/g' '$(PREFIX)/$(TARGET)/x86/mingw/lib/OpenCVConfig.cmake')
```
Comment (TimothyGu) [inline src/opencv.mk]: ditto

File: src/netcdf.mk [hunk 19]
```diff
@@ -26,20 +26,16 @@ define $(PKG)_BUILD
     cd '$(1)' && \
         $(SED) -i -e 's/#ifdef IGNORE/#if 0/' libsrc4/nc4hdf.c libsrc4/ncfunc.c libsrc/attr.c ncgen/cvt.c && \
         ./configure \
-        --host='$(TARGET)' \
-        --build="`config.guess`" \
-        --disable-shared \
+    $(MXE_CONFIGURE_OPTS) \
```
Comment (TimothyGu) [inline src/netcdf.mk]: indentation, here and below

File: src/netcdf.mk [hunk 20]
```diff
@@ -26,20 +26,16 @@ define $(PKG)_BUILD
...
         LIBS="-lmfhdf -ldf -lportablexdr -lws2_32"
-    $(MAKE) -C '$(1)' -j '$(JOBS)'
-    $(MAKE) -C '$(1)' -j 1 install
-endef
 
-$(PKG)_BUILD_x86_64-w64-mingw32 =
-$(PKG)_BUILD_i686-w64-mingw32 = $(subst --enable-hdf4, --disable-hdf4,\
-                                $(subst -lmfhdf -ldf,,$($(PKG)_BUILD)))
+    $(MAKE) -C '$(1)' -j '$(JOBS)' \
+    LDFLAGS=-no-undefined
```
Comment (TimothyGu) [inline src/netcdf.mk]: You can coalesce these two lines

File: src/muparserx.mk [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+$(PKG)_IGNORE   :=
+$(PKG)_VERSION  := v3_0_3
+$(PKG)_CHECKSUM := d476899a024b6b720591484f615464dc1eb25b23
+$(PKG)_SUBDIR   :=
+$(PKG)_FILE     := $(PKG)_v3_0_2.zip
+$(PKG)_URL      := http://muparserx.googlecode.com/svn/archives/$($(PKG)_FILE)
+$(PKG)_DEPS     := gcc
+
+define $(PKG)_UPDATE
+echo "TODO: write update for muparserx"
```
Comment (TimothyGu) [inline src/muparserx.mk]: indentation

File: src/muparserx.mk [hunk 22]
```diff
@@ -0,0 +1,25 @@
+# This file is part of MXE.
+# See index.html for further information.
+
+PKG             := muparserx
+$(PKG)_IGNORE   :=
+$(PKG)_VERSION  := v3_0_3
+$(PKG)_CHECKSUM := d476899a024b6b720591484f615464dc1eb25b23
+$(PKG)_SUBDIR   :=
+$(PKG)_FILE     := $(PKG)_v3_0_2.zip
```
Comment (TimothyGu) [inline src/muparserx.mk]: is `v3_0_2` right? You can use `$($(PKG)_VERSION)`

File: src/muparserx.mk [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+define $(PKG)_UPDATE
+echo "TODO: write update for muparserx"
+endef
+
+define $(PKG)_BUILD
+    mkdir '$(1)$(PKG)_$($(PKG)_VERSION).build'
+    cd '$(1)$(PKG)_$($(PKG)_VERSION).build' && cmake \
+        -DCMAKE_TOOLCHAIN_FILE='$(CMAKE_TOOLCHAIN_FILE)' \
+        -DBUILD_SHARED_LIBS=$(if $(BUILD_STATIC),FALSE,TRUE) \
+    -DCMAKE_CXX_FLAGS='-std=c++11' \
```
Comment (TimothyGu) [inline src/muparserx.mk]: broken indentation

File: src/muparserx.mk [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+$(PKG)_FILE     := $(PKG)_v3_0_2.zip
+$(PKG)_URL      := http://muparserx.googlecode.com/svn/archives/$($(PKG)_FILE)
+$(PKG)_DEPS     := gcc
+
+define $(PKG)_UPDATE
+echo "TODO: write update for muparserx"
+endef
+
+define $(PKG)_BUILD
+    mkdir '$(1)$(PKG)_$($(PKG)_VERSION).build'
```
Comment (TimothyGu) [inline src/muparserx.mk]: here and below, I'd rather put a `/` between `$(1)` and `$(PKG)_$($(PKG)_VERSION)`

File: src/muparserx-1-cmakelists.patch [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+--- patch_test/muparserx_v3_0_3.orig/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
++++ patch_test/muparserx_v3_0_3/CMakeLists.txt	2015-02-23 20:25:36.710543187 +0100
+@@ -0,0 +1,14 @@
++cmake_minimum_required(VERSION 2.8)
++file(GLOB MUPARSERX_SRCS "${CMAKE_SOURCE_DIR}/parser/*.cpp" )
++
++add_library(muparserx ${MUPARSERX_SRCS})
++
++install(TARGETS muparserx
++  RUNTIME DESTINATION bin COMPONENT RuntimeLibraries
```
Comment (TimothyGu) [inline src/muparserx-1-cmakelists.patch]: you don't strictly need `COMPONENT`. Here and below, and in other files too.

File: src/muparser.mk [hunk 26]
```diff
@@ -20,10 +20,8 @@ define $(PKG)_BUILD
     cd '$(1)' && ./configure \
         --host='$(TARGET)' \
         --prefix='$(PREFIX)/$(TARGET)' \
-        --disable-shared \
+        $(if $(BUILD_STATIC),--enable-static,--enable-shared) \
```
Comment (TimothyGu) [inline src/muparser.mk]: `$(MXE_CONFIGURE_OPTS)`

File: src/libgeotiff-2-libtoolfilemagic.patch [hunk 27]
```diff
@@ -0,0 +1,12 @@
+diff -bur libgeotiff-1.4.0.orig/ltmain.sh libgeotiff-1.4.0/ltmain.sh
+--- libgeotiff-1.4.0.orig/ltmain.sh	2014-10-24 01:20:03.172985419 +0530
++++ libgeotiff-1.4.0/ltmain.sh	2014-10-24 01:58:50.983491081 +0530
+@@ -2560,7 +2560,7 @@
+     ;;
+   *ar\ archive*) # could be an import, or static
+     if eval $OBJDUMP -f $1 | $SED -e '10q' 2>/dev/null |
+-       $EGREP 'file format pe-i386(.*architecture: i386)?' >/dev/null ; then
++       $EGREP 'file format pe-i386(.*architecture: i386)?|pe-x86-64' >/dev/null ; then
```
Comment (TimothyGu) [inline src/libgeotiff-2-libtoolfilemagic.patch]: `ltmain.sh` is generated automatically; you should run `autoreconf -fi` in the recipe file to update this file instead of modifying it by hand.

File: src/itk.mk [hunk 28]
```diff
@@ -20,13 +21,31 @@ define $(PKG)_BUILD
...
+        -DVXL_HAS_SSE2_HARDWARE_SUPPORT=1 \
+        -DKWSYS_CHAR_IS_SIGNED=1 \
+        -DKWSYS_LFS_WORKS=1 \
         '$(1)'
-    $(MAKE) -C '$(1).build' -j '$(JOBS)' install VERBOSE=1
-endef
+    # make
+    $(MAKE) -C '$(1).build' -j '$(JOBS)'
+    # install
+    $(MAKE) -C '$(1).build' -j 1 install #VERBOSE=1
```
Comment (TimothyGu) [inline src/itk.mk]: You can leave these two lines alone. They were fine before.

File: src/gdal-2-openjpeg2.1.0rev27538.patch [hunk 29]
```diff
@@ -0,0 +1,247 @@
...
++    fi
+   fi
+ else
+ 
+   HAVE_OPENJPEG=yes
+   if test -r $with_openjpeg/include/openjpeg-2.0/openjpeg.h ; then
+     EXTRA_INCLUDES="-I$with_openjpeg/include $EXTRA_INCLUDES"
++  elif test -r $with_openjpeg/include/openjpeg-2.1/openjpeg.h ; then
++    OPENJPEG_VERSION=20100
++    EXTRA_INCLUDES="-I$with_openjpeg/include $EXTRA_INCLUDES"
```
Comment (TimothyGu) [inline src/gdal-2-openjpeg2.1.0rev27538.patch]: I'd put an extra pair of quotation marks around `'$with_openjpeg/include'`

Comment (TimothyGu) [inline src/gdal-2-openjpeg2.1.0rev27538.patch]: scratch that

File: src/gdal-2-openjpeg2.1.0rev27538.patch [hunk 30]
```diff
@@ -0,0 +1,247 @@
+diff -bur gdal-1.11.1.orig/configure gdal-1.11.1/configure
+--- gdal-1.11.1.orig/configure	2014-10-24 10:57:41.487491392 +0530
++++ gdal-1.11.1/configure	2014-10-24 10:58:16.354947432 +0530
```
Comment (TimothyGu) [inline src/gdal-2-openjpeg2.1.0rev27538.patch]: Leave this file alone, and `autoreconf -fi` in the script

Comment (TimothyGu) [inline src/gdal-2-openjpeg2.1.0rev27538.patch]: Scratch that.

File: src/gdal.mk [hunk 31]
```diff
@@ -23,17 +23,12 @@ define $(PKG)_CONFIGURE
     cd '$(1)' && autoreconf -fi
     # The option '--without-threads' means native win32 threading without pthread.
     cd '$(1)' && ./configure \
-        --host='$(TARGET)' \
-        --build='$(BUILD)' \
-        --enable-static \
-        --disable-shared \
-        --prefix='$(PREFIX)/$(TARGET)' \
+    $(MXE_CONFIGURE_OPTS) \
```
Comment (TimothyGu) [inline src/gdal.mk]: indentation

File: src/gdal.mk [hunk 32]
```diff
@@ -45,8 +40,11 @@ define $(PKG)_CONFIGURE
         --with-sqlite3='$(PREFIX)/$(TARGET)' \
         --with-gta='$(PREFIX)/$(TARGET)' \
         --with-hdf5='$(PREFIX)/$(TARGET)' \
+        --with-netcdf='$(PREFIX)/$(TARGET)' \
+    --with-openjpeg='$(PREFIX)/$(TARGET)' \
```
Comment (TimothyGu) [inline src/gdal.mk]: Ditto

File: src/gdal.mk [hunk 33]
```diff
@@ -82,8 +80,9 @@ define $(PKG)_MAKE
     $(MAKE) -C '$(1)/gcore' -j '$(JOBS)' install
     $(MAKE) -C '$(1)/frmts' -j '$(JOBS)' install
     $(MAKE) -C '$(1)/alg'   -j '$(JOBS)' install
-    $(MAKE) -C '$(1)/ogr'   -j '$(JOBS)' install OGR_ENABLED=
-    $(MAKE) -C '$(1)/apps'  -j '$(JOBS)' install BIN_LIST=
+    $(MAKE) -C '$(1)/ogr'   -j '$(JOBS)' install #OGR_ENABLED=
+    $(MAKE) -C '$(1)/apps'  -j '$(JOBS)' install
+    $(MAKE) -C '$(1)'       -j '$(JOBS)' install #make install on each dir is required?
```
Comment (TimothyGu) [inline src/gdal.mk]: It is not required from a quick look, but the original intention is not to install every single directory. Also, leave the `BIN_LIST=` alone as MXE doesn't build applications.

File: src/hdf4.mk [hunk 34]
```diff
@@ -21,19 +21,22 @@ define $(PKG)_BUILD
...
-        --prefix='$(PREFIX)/$(TARGET)' \
-        CPPFLAGS="-DH4_F77_FUNC\(name,NAME\)=NAME -DH4_BUILT_AS_STATIC_LIB=1"
-    $(MAKE) -C '$(1)'/hdf/src -j '$(JOBS)'
+    $(if $(BUILD_STATIC), \
+        CPPFLAGS="-DH4_F77_FUNC\(name,NAME\)=NAME -DH4_BUILT_AS_STATIC_LIB=1") \
+        AR='$(TARGET)-ar' \
+    $(if $(BUILD_SHARED), \
+        LIBS="-lportablexdr -lws2_32" CPPFLAGS="-DH4_F77_FUNC\(name,NAME\)=NAME -DH4_BUILT_AS_DYNAMIC_LIB=1 -DBIG_LONGS")
+    $(MAKE) -C '$(1)'/mfhdf/xdr -j '$(JOBS)' \
+    LDFLAGS=-no-undefined
```
Comment (TimothyGu) [inline src/hdf4.mk]: Please fix the indentation of `LDFLAGS` here and below.

File: src/itk-3-execvp.patch [hunk 35]
```diff
@@ -0,0 +1,12 @@
+diff -burN ITK.original/Modules/ThirdParty/KWSys/src/KWSys/SharedForward.h.in ITK/Modules/ThirdParty/KWSys/src/KWSys/SharedForward.h.in
+--- ITK.original/Modules/ThirdParty/KWSys/src/KWSys/SharedForward.h.in	2014-07-30 03:38:46.547938233 +0200
++++ ITK/Modules/ThirdParty/KWSys/src/KWSys/SharedForward.h.in	2014-07-30 03:37:16.114246728 +0200
+@@ -512,6 +512,8 @@
+   /* Invoke the child process.  */
+ #if defined(_MSC_VER)
+   _execvp(cmd, argv);
++#elif defined(__MINGW64__) || defined(__MINGW32__)
```
Comment (TimothyGu) [inline src/itk-3-execvp.patch]: The second one only should suffice.

File: src/itk.mk [hunk 36]
```diff
@@ -3,16 +3,17 @@
...
 $(PKG)_FILE     := $($(PKG)_SUBDIR).tar.xz
 $(PKG)_URL      := http://$(SOURCEFORGE_MIRROR)/project/$(PKG)/$(PKG)/$(call SHORT_PKG_VERSION,$(PKG))/$($(PKG)_FILE)
-$(PKG)_DEPS     := gcc hdf5
+$(PKG)_DEPS     := gcc hdf5 libpng tiff jpeg expat
 
 define $(PKG)_UPDATE
-    echo 'TODO: Updates for package ITK need to be written.' >&2;
-    echo $(itk_VERSION)
+    $(WGET) -q -O- 'https://sourceforge.net/projects/itk/files/itk/' | \
+    $(SED) -n 's,.*/\([0-9][^"]*\)/".*,\1,p' | \
```
Comment (TimothyGu) [inline src/itk.mk]: I suspect a stricter check might be needed (one with `\.tar\..*`)

Comment (TimothyGu) [inline src/itk.mk]: Sorry ignore this comment please

File: index.html (hunk 1 above)
Comment (rashadkm) [inline index.html]: why do you say this isnt ossim?

File: src/qwt5_qt4-1-exampleslib.patch (hunk 6 above)
Comment (rashadkm) [inline src/qwt5_qt4-1-exampleslib.patch]: any suggestions?

File: src/otb.mk (hunk 11 above)
Comment (rashadkm) [inline src/otb.mk]: these are shared libs loaded by otbApplicationLauncherCommandLine. for the ease there are .sh and .bat scripts for every application

File: src/otb.mk (hunk 12 above)
Comment (rashadkm) [inline src/otb.mk]: didnt get that

File: src/otb.mk (hunk 13 above)
Comment (rashadkm) [inline src/otb.mk]: no internal curl here. but if you dont have curl certain part of code are not included via ifndef

File: src/opencv.mk (hunk 15 above)
Comment (rashadkm) [inline src/opencv.mk]: opencv is using cmake. if you need proper verbose that will be CMAKE_VERBOSE_MAKEFILE=TRUE. 

maybe could be adding in mxe toolchain. This is only for if one choose to actually have verbose.

File: src/otb.mk (hunk 10 above)
Comment (rashadkm) [inline src/otb.mk]: examples and testing takes too much time. Although I love to turn it TRUE. and for PTHREADS vs WIN32THREADS, do you have any suggestions or should I leave it to cmake default ?

File: src/opencv.mk (hunk 18 above)
Comment (rashadkm) [inline src/opencv.mk]: no one is for x86 and other is x64

File: src/itk-3-execvp.patch (hunk 35 above)
Comment (rashadkm) [inline src/itk-3-execvp.patch]: not clear

**Comment by rashadkm** [comment]
what is the way to check/fix intention error ?

File: index.html (hunk 1 above)
Comment (TimothyGu) [inline index.html]: you put `ossim` as the content of `<a>`

File: src/otb.mk (hunk 10 above)
Comment (TimothyGu) [inline src/otb.mk]: i'd enable WIN32THREADS

File: src/otb.mk (hunk 12 above)
Comment (TimothyGu) [inline src/otb.mk]: Why is `OTB_USE_PATENTED` is `FALSE`?

File: src/otb.mk (hunk 13 above)
Comment (TimothyGu) [inline src/otb.mk]: cool

File: src/itk-3-execvp.patch (hunk 35 above)
Comment (TimothyGu) [inline src/itk-3-execvp.patch]:
```c
#elif defined(__MINGW32__)
```

is fine.

`__MINGW32__` is defined on 64-bit systems as well.

**Comment by TimothyGu** [comment]
Add spaces.

**Comment by rashadkm** [comment]
here are files that need fix:
src/libgeotiff-2-libtoolfilemagic.patch
src/muparserx-1-cmakelists.patch
src/gdal-2-openjpeg2.1.0rev27538.patch
src/muparserx.mk
src/portablexdr.mk
src/qwt5_qt4-1-exampleslib.patch

Are we clear upto here?

File: src/gdal-2-openjpeg2.1.0rev27538.patch (hunk 30 above)
Comment (rashadkm) [inline src/gdal-2-openjpeg2.1.0rev27538.patch]: @TimothyGu, could you tell about autoreconf -fi, using that I dont need to modify configure script is that right?

**Comment by saiarcot895** [comment]
Is there any chance this will be merged in soon?

**Comment by rashadkm** [comment]
@saiarcot895 , read this thread regarding discussion of this work  - http://lists.nongnu.org/archive/html/mingw-cross-env-list/2015-05/msg00015.html

Long story, short.. the pull request had grown too much and I was suggested to open multiple pull requests. So that it easy to review and merge
